api/base/authentication/drf.py

```diff
@@ -1,88 +1,166 @@
-    """Checks that the user is neither unconfirmed nor disabled.
-
-    :param User user:
-    :raises DeactivatedAccountError:
-    :raise UnconfirmedAccountError:
     """
+    Verify users' status.
+
+                        registered      confirmed       disabled        merged      usable password     claimed
+    ACTIVE:             x               x               o               o           x                   x
+    NOT_CONFIRMED:      o               o               o               o           x                   o
+    NOT_CLAIMED:        o               o               o               o           o                   o
+    DISABLED:           o               x               x               o           x                   x
+    USER_MERGED:        x               x               o               x           o                   x
+
+    :param user: the user
+    :raises UnconfirmedAccountError
+    :raises UnclaimedAccountError
+    :raises DeactivatedAccountError
+    :raises MergedAccountError
+    :raises InvalidAccountError
+    """
+
+    # active user must be registered, claimed, confirmed, not merged or disabled, and has a usable password
+    if user.is_active:
+        return
+
+    # user disabled
     if user.is_disabled:
         raise DeactivatedAccountError
-    elif not user.is_confirmed:
-        raise UnconfirmedAccountError
+
+    # user merged
+    if user.is_merged:
+        raise MergedAccountError
+
+    # user not confimred or contributor not claimed
+    if not user.is_confirmed and not user.is_registered and not user.is_claimed:
+        if user.password is None or user.password.startswith(UNUSABLE_PASSWORD_PREFIX):
+            raise UnclaimedAccountError
+        if user.has_usable_password():
+            raise UnconfirmedAccountError
+
+    # OSF does not recognize other user status.
+    raise InvalidAccountError
+
+
+# Three customized DRF authentication classes: basic, session/cookie and access token.
+# See http://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
 
 
-# http://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
 class OSFSessionAuthentication(authentication.BaseAuthentication):
-    """Custom DRF authentication class which works with the OSF's Session object.
+    """
+    Custom DRF authentication class for API call with OSF cookie/session.
     """
 
     def authenticate(self, request):
+        """
+        If request bears an OSF cookie, retrieve the session and verify the user.
+
+        :param request: the request
+        :return: the user
+        """
         cookie_val = request.COOKIES.get(settings.COOKIE_NAME)
         if not cookie_val:
             return None
         session = get_session_from_cookie(cookie_val)
         if not session:
             return None
         user_id = session.data.get('auth_user_id')
-        user = User.load(user_id)
+        user = OSFUser.objects.filter(guids___id=user_id).first()
```

Comment:
`OSFUser.objects.get(guids___id=user_id)` (catch the exception if needed, `OSFUser.ObjectDoesNotExist`)